### PR TITLE
Fix analyze issues

### LINKED ISSUE #1048: The .NET Upgrade Assistant tool should provide guidance for each warning logged to make them actionable
https://developercommunity.visualstudio.com/t/The-NET-Upgrade-Assistant-tool-should-p/1657378

When using the .NET Upgrade Assistant “DotNet” tool version 0.3.310801 to analyze a C# project, I get a SARIF file that I can open in Visual Studio using the Tools -> Open Static Analysis Results as SARIF and it shows a list of warnings in the Error List window. In my case it shows 1598 warnings and I can use the column filters to select specific projects:

image

What I am missing is guidance on what I do about these warnings. When I click the UAXXX link in the Code column, it takes me to https://docs.microsoft.com/en-us/dotnet/core/porting/upgrade-assistant-overview for all of them, but there is no mentioning of what the UA code means.

In the screenshot, it says “UA105 - Unnecessary transitive dependency”. It does not tell from which package or project it finds a transitive dependency that is not needed and it does not tell me the name and version of the transitive dependency it deemed unnecessary or why it is unnecessary. Thus, I do not know what I am supposed to do with this information.

Similarly, the warning “UA108 - Framework Reference to Microsoft.AspNetCore.App needs to be added” does not tell me why it needs to be added or where to add it.

I also get “UA106 - Package Microsoft.Net.Http needs to be upgraded from 2.2.22 to 2.2.29” warnings, but no information why.

image

The above warnings are all from the same csproj file, so there seem to be multiple duplicate warnings for the same project and many do not have a line number, so clicking click the message may open the project, but it does not go to a specific line.

I like that I can now run the tool on my projects using centrally managed package versions and get a list of issues. What I now need is guidance on these issues - what do they each mean, why is an issue preventing me from upgrading to a newer .NET version, and what action should I take to solve it.

### PR DESCRIPTION
**Pull request by sunandabalu**
- [x] Add more details to diagnostic messages for Package Dependency analysis for better user experience
- [x] Fix bug where some of the diagnostic messages where repeating for the same package in the same .csproj file.
- [x] Notify user when the analysis is complete along with where the analysis report can be found

Fixes #1048
#1047

### CODE REVIEW

File: src/extensions/default/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/Analyzers/TargetCompatibilityReferenceAnalyzer.cs [hunk 1]
```diff
@@ -65,26 +65,27 @@ public async Task AnalyzeAsync(IProject project, IDependencyAnalysisState state,
...
+                        var logMessage = SR.Format("Package {0} does not support the target(s) {1} but a newer version ({2}) does.", packageReference, string.Join(", ", state.TargetFrameworks), updatedReference.Version);
+                        _logger.LogInformation(logMessage);
                         var isMajorChange = _comparer.IsMajorChange(updatedReference.Version, packageReference.Version);
 
                         if (isMajorChange)
                         {
-                            var logString = SR.Format("Package {0} needs to be upgraded across major versions ({1} -> {2}) which may introduce breaking changes", packageReference.Name, packageReference.Version, updatedReference.Version);
-                            details.Add(logString);
+                            var logString = SR.Format("Package {0} needs to be upgraded across major versions ({1} -> {2}) which may introduce breaking changes.", packageReference.Name, packageReference.Version, updatedReference.Version);
+                            details.Add(string.Concat(logMessage, logString));
```
Comment (brandonh-msft) [inline src/extensions/default/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/Analyzers/TargetCompatibilityReferenceAnalyzer.cs:75]: Do you want to put a <space> between the two? Maybe interpolation would work?

File: src/extensions/default/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/Analyzers/DuplicateReferenceAnalyzer.cs [hunk 2]
```diff
@@ -47,8 +47,9 @@ public Task AnalyzeAsync(IProject project, IDependencyAnalysisState state, Cance
 
                 foreach (var package in duplicates.Where(p => p != highestVersion))
                 {
-                    _logger.LogInformation("Marking package {NuGetPackage} for removal because it is referenced elsewhere in the project with a higher version", package);
-                    state.Packages.Remove(package, new OperationDetails());
+                    var logMessage = SR.Format("Marking package {0} for removal because it is referenced elsewhere in the project with a higher version", package);
```
[the remark this answers is not in the document]
Comment (brandonh-msft) [reply src/extensions/default/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/Analyzers/DuplicateReferenceAnalyzer.cs:50]: yup decided to clone the PR and check it out. 👍🏻

File: src/extensions/default/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/Analyzers/TargetCompatibilityReferenceAnalyzer.cs (hunk 1 above)
Comment (sunandabalu) [reply src/extensions/default/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/Analyzers/TargetCompatibilityReferenceAnalyzer.cs]: Good point, changed to use interpolation.

File: src/extensions/default/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/PackageAnalyzerResultProvider.cs [hunk 3]
```diff
@@ -150,30 +150,17 @@ void GetResults<T>(string ruleId, string name, string action, IReadOnlyCollectio
...
-                                    RuleId = ruleId,
-                                    RuleName = ruleName,
-                                    FullDescription = fullDescription,
-                                    HelpUri = _helpUri,
-                                    FileLocation = fileLocation,
-                                    ResultMessage = s,
-                                }));
-                            }
-                            else
+                            var resultMessage = (s.OperationDetails is not null && s.OperationDetails.Details is not null && s.OperationDetails.Details.Any() && !results.Select(i => i.FileLocation == fileLocation & i.ResultMessage == s.OperationDetails.Details.FirstOrDefault()).Any()) ? s.OperationDetails.Details.FirstOrDefault() : string.Concat(name, s.Item, action);
```
[the remark this answers is not in the document]
Comment (sunandabalu) [reply src/extensions/default/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/PackageAnalyzerResultProvider.cs]: Changed

**Review by brandonh-msft — APPROVED** [review]

**Pull request merged**